### ISSUE TITLE
attempt to concatenate field 'state' (a nil value)

### ISSUE THREAD

**Issue by consultantRR** [pingidentity/lua-resty-openidc#3]
This is working very well, except on about 10-15% of page loads (two different sites) typically after an initial successful login to the site, I am getting 500's which show up as `attempt to concatenate field 'state' (a nil value)` in the logs.

Here are two examples:

```
2016/01/10 08:41:18 [error] 20680#0: *198 lua entry thread aborted: runtime error: /opt/openresty/lualib/resty/openidc.lua:260: attempt to concatenate field 'state' (a nil value)
stack traceback:
coroutine 0:
	/opt/openresty/lualib/resty/openidc.lua: in function 'authenticate'
	access_by_lua(nginx.conf:232):11: in function <access_by_lua(nginx.conf:232):1>, client: IP_REDACTED, server: REDACTED.com, request: "GET /redirect_uri?state=1f3a05983ecad4f7c567cec9906ec613&code=4/xfw5ux-IejCSEcojEcgs-fcxqz1J6V5xF2lORuipuFw&authuser=0&hd=REDACTED.com&session_state=d9e845af3ee7ece2958d2c45dc3abf123b77a056..d752&prompt=none HTTP/1.1", host: "REDACTED.com", referrer: "https://REDACTED.com/jsp/common/login/Login.jsp;jsessionid=G1gFWSXVrTylsXdHkRTp9hRcmXKcyTfZWF8rzsmGLmpk76TwDlvj!-1484896595"
```

```
2016/01/10 07:51:27 [error] 20526#0: *14 lua entry thread aborted: runtime error: /opt/openresty/lualib/resty/openidc.lua:260: attempt to concatenate field 'state' (a nil value)
stack traceback:
coroutine 0:
	/opt/openresty/lualib/resty/openidc.lua: in function 'authenticate'
	access_by_lua(nginx.conf:188):11: in function <access_by_lua(nginx.conf:188):1>, client: IP_REDACTED, server: REDACTED.com, request: "GET /redirect_uri?state=4ff8e3db1e9490dd2e5c9e9106a211b5&code=4/W8OlE5WneQJRJ37QmKWUH_XO3EoCfGxDOKJ1FoVqZ-Y&authuser=0&hd=REDACTED.com&session_state=85ec1da20441f46f50edc2dd7b190bd96b7e4b4d..d98b&prompt=none HTTP/1.1", host: "REDACTED.com", referrer: "https://accounts.google.com/ServiceLogin?continue=https%3A%2F%2Faccounts.google.com%2Fo%2Foauth2%2Fv2%2Fauth%3Fscope%3Dopenid%2Bemail%2Bprofile%26response_type%3Dcode%26state%3D4ff8e3db1e8690dd2e5c9e9106a211b5%26redirect_uri%3Dhttps%3A%2F%2FREDACTED.com%2Fredirect_uri%26nonce%3Db58d3254d725632575716345f28b1a8b%26client_id%3DREDACTED.apps.googleusercontent.com%26hl%3Den%26from_login%3D1%26as%3D7e62210aa03c7ffe&ltmpl=popup&sacu=1&scc=1&passive=1209600&acui=0"
```

**Comment by zandbelt** [comment]
diff --git a/lib/resty/openidc.lua b/lib/resty/openidc.lua
    index ee3fe83..cca46ec 100644
    --- a/lib/resty/openidc.lua
    +++ b/lib/resty/openidc.lua
    @@ -257,7 +257,7 @@
you may apply this patch to prevent the null pointer exception but something else is wrong: it does not seem that the session is begin properly restored     

       -- check that the state returned in the response against the session; prevents CSRF
       if args.state ~= session.data.state then
    -    err = "state from argument: "..args.state.." does not match state restored from session:     "..session.data.state
    +    err = "state from argument: "..args.state and args.state or "nil".." does not match state restored     from session: "..session.data.state and session.data.state or "nil"
         ngx.log(ngx.ERR, err)
         return nil, err
       end

**Comment by zandbelt** [comment]
can you check if it has to do with: https://github.com/bungle/lua-resty-session/issues/9

**Comment by consultantRR** [comment]
Hmm, it does seem possible. Let me disable `session_check_ssi` and test a bit. (I am just diving into Lua today because of your oidc.) So if you are easily able to post syntax to add to my nginx.conf to do this in the right context and make sure my test is valid, that would be great. Otherwise, I'll figure it out, and report back after testing.

**Comment by consultantRR** [comment]
Disabling `session_check_ssi` does seem to have helped. I will do some further testing and report back tomorrow.

**Comment by bungle** [comment]
Thank you for this, I may reconsider setting session_check_ssi as `false` by default (it doesn't do anything with TLS after all). It could be that SSL session has timeouted if it timeouts differently than session: https://github.com/bungle/lua-resty-session#nginx-configuration-variables

**Comment by consultantRR** [comment]
Just reporting back that, indeed, after disabling `session_check_ssi` I did not see this issue recur. I do occasionally see some odd ssl messages in the logs and I am trying to find out if they are in anyway related to this. If so, I will post here. If not, I won't pollute the thread!

**Comment by zandbelt** [comment]
I'm going to close this issue as (after fixing the nil pointer exception) it is not specific to `lua-resty-openidc`; feel free to add comments if you find out more about the related issues in `lua-resty-session`.

**Comment by bungle** [comment]
I just removed released 2.12 of lua-resty-session that does disable session_check_ssi by default.
